web-api/src/storage/memory.rs

```diff
@@ -321,50 +323,49 @@ impl storage::Document for Storage {
         Ok(documents)
     }
 
-    async fn insert(
-        &self,
-        documents_embeddings: Vec<(IngestedDocument, NormalizedEmbedding)>,
-    ) -> Result<(), InsertionError> {
-        if documents_embeddings.is_empty() {
-            return Ok(());
+    async fn insert(&self, new_documents: Vec<IngestedDocument>) -> Result<Vec<DocumentId>, Error> {
+        if new_documents.is_empty() {
+            return Ok(Vec::new());
         }
 
         let mut documents = self.documents.write().await;
         let mut embeddings = mem::take(&mut documents.1).into_heads().map;
-        documents.0.reserve(documents_embeddings.len());
-        for (document, embedding) in documents_embeddings {
+        documents.0.reserve(new_documents.len());
+        for document in new_documents {
             documents.0.insert(
                 document.id.clone(),
                 Document {
                     properties: document.properties,
                     tags: document.tags,
+                    #[cfg(feature = "ET-4089")]
+                    is_candidate: document.is_candidate,
                 },
             );
-            embeddings.insert(document.id, embedding);
+            embeddings.insert(document.id, document.embedding);
         }
         documents.1 = Embeddings::borrowed(embeddings);
 
-        Ok(())
+        Ok(Vec::new())
     }
 
     async fn delete(
         &self,
         ids: impl IntoIterator<IntoIter = impl Clone + ExactSizeIterator<Item = &DocumentId>>,
-    ) -> Result<(), DeletionError> {
+    ) -> Result<Vec<DocumentId>, Error> {
         let mut documents = self.documents.write().await;
         let mut interactions = self.interactions.write().await;
 
-        let ids = ids.into_iter().collect::<HashSet<_>>();
-        documents.0.retain(|id, _| !ids.contains(id));
-        let mut embeddings = mem::take(&mut documents.1).into_heads().map;
-        embeddings.retain(|id, _| !ids.contains(id));
-        documents.1 = Embeddings::borrowed(embeddings);
+        let mut ids = ids.into_iter().collect::<HashSet<_>>();
         interactions.retain(|_, interactions| {
             interactions.retain(|(id, _)| !ids.contains(id));
             !interactions.is_empty()
         });
+        documents.0.retain(|id, _| !ids.contains(id));
+        let mut embeddings = mem::take(&mut documents.1).into_heads().map;
+        embeddings.retain(|id, _| !ids.remove(id));
```

Comment:
:wrench: 

```suggestion
        embeddings.retain(|id, _| !ids.contains(id));
```

If we anyway drop `ids` a moment later we should use `contains` here I think as it more represents what we do and earlier removing is not an optimization either. Or did I miss something?

Reply:
we don't drop them, but return the remaining ids a few lines later. the leftover ids are the ones which are not present in the db but were requested to be deleted, which aligns it with the es/pg db impl.